agera/src/main/java/com/google/android/agera/FunctionCompiler.java

```diff
@@ -197,52 +190,99 @@ public Function thenSort(@NonNull final Comparator comparator) {
     @SuppressWarnings("unchecked")
     public Object apply(@NonNull final Object input) {
       Object item = input;
-      for (final Function function : functions) {
-        item = function.apply(item);
-      }
-      return item;
-    }
-  }
-
-  private static final class FilterFunction<T> implements Function<List<T>, List<T>> {
-    @NonNull
-    private final Predicate filter;
+      boolean compilerList = false;
+      int i = 0;
+      final int length = functions.length;
+      while (i < length) {
+        final int type = (int) functions[i++];
+        final Object command = functions[i++];
+        switch (type) {
+          case LIMIT: {
+            final List list = (List) item;
+            final int limit = (int) command;
+            if (limit <= 0) {
+              item = emptyList();
+            } else if (list.size() < limit) {
+              item = input;
+            } else if (compilerList) {
+              while (list.size() > limit) {
```

Comment:
Actually, looked into this and it allocates too? :P I'm looking into replacing the equivalent line under FILTER, hang on.